1.1/MetadataGranularities/README.md

```diff
@@ -0,0 +1,200 @@
+
+## Screenshot
+
+![MetadataGranularities](screenshot/MetadataGranularities.png)
+
+## Sandcastle Code
+
+```JavaScript
+var viewer = new Cesium.Viewer('cesiumContainer');
+
+// Create the tileset, and set its model matrix to move it
+// to a certain position on the globe
+var tileset = viewer.scene.primitives.add(new Cesium.Cesium3DTileset({
+    url : 'http://localhost:8003/1.1/MetadataGranularities/tileset.json',
+    maximumScreenSpaceError: 1,
+    debugShowBoundingVolume: true
+}));
+tileset.modelMatrix = Cesium.Transforms.eastNorthUpToFixedFrame(
+  Cesium.Cartesian3.fromDegrees(-75.152325, 39.94704, 0.0));
+
+// Create the label that will display metadata information
+var labelEntity = viewer.entities.add({
+  label: {
+    showBackground: true,
+    font: "14px monospace",
+    horizontalOrigin: Cesium.HorizontalOrigin.LEFT,
+    verticalOrigin: Cesium.VerticalOrigin.TOP,
+    pixelOffset: new Cesium.Cartesian2(15, 0),
+    // Workaround to always render label on top:
+    disableDepthTestDistance: Number.POSITIVE_INFINITY
+  }
+});
+
+// Adds metadata information to the label, with 
+// the given title indicating which granularity
+// of metadata this is. The given metadata object
+// can be a TilesetMetadata, TileMetadata, 
+// GroupMetadata, or ContentMetadata.
+var addMetadataToLabel = function(title, metadata) {
+  if (!Cesium.defined(metadata)) {
+    labelEntity.label.text += "(No " + title + ")\n";
+    return;
+  }
+  var propertyKeys = metadata.getPropertyIds();
+  if (!Cesium.defined(propertyKeys)) {
+    labelEntity.label.text += "(No properties for " + title + ")\n";
+  }
+  labelEntity.label.text += title + ":\n";
+  for (var i=0; i<propertyKeys.length; i++) {
+    var propertyKey = propertyKeys[i];
+    var propertyValue = metadata.getProperty(propertyKey);          
+    //console.log("propertyKey", propertyKey);
+    //console.log("propertyValue", propertyValue);
+    labelEntity.label.text += "  " + propertyKey + ": "+ propertyValue;
+    labelEntity.label.text += "\n";
+  }
+};
+
+// (NOTE: The way how metadata is obtained may be refactored,
+// see https://github.com/CesiumGS/cesium/issues/10015 )
+
+// Given an object that was obtained via Scene#pick, examine it 
+// to see whether it contains tile metadata.
+// If the object contains TileMetadata, then this metadata 
+// is returned. 
+// Otherwise, 'undefined' is returned.
+var obtainTileMetadata = function(picked) {
+  if (!Cesium.defined(picked)) {
+    return undefined;
+  }
+  if (!Cesium.defined(picked.content)) {
+    return undefined;
+  }
+  if (!Cesium.defined(picked.content.tile)) {
+    return undefined;
+  }
+  if (!Cesium.defined(picked.content.tile.metadata)) {
+    return undefined;
+  }
+  var metadata = picked.content.tile.metadata;
+  var isTileMetadata = metadata instanceof Cesium.TileMetadata;
+  if (!isTileMetadata) {
+    return undefined;
+  }
```

Comment:
When would it be a type other than `Cesium.TileMetadata`? Could the check simply be whether it's defined or not?

Also.. thinking about future demos.. what does this look like for implicit tiling? Is it the same API or different? CC @j9liu

Reply:
`tile.metadata` can be of `Cesium.TileMetadata` (explicit) or `Cesium.ImplicitMetadataView`. However, the APIs for both are the same, so I think the check is not necessary. Along those lines, because the API for implicit tiling metadata should be the same, this sandcastle should function similarly for an implicit tileset.

Reply:
The `ImplicitMetadataView` was once called `ImplicitTileMetadata`, and I had figured that out in https://github.com/CesiumGS/3d-tiles-samples/blob/a546b82299ae5700525ba825564f857130a4e25f/EXT_mesh_features/CommonTestingSandcastleCode.txt#L119 , but didn't handle it here, because there are no implicit tilesets in this example. 

The general question of how rigurously all this should be checked remains open for me. E.g. I think that Peter once mentioned that ~"If _this_ is present, then _that_ is also present" (in view of the somewhat tedious `defined(...)` checks). But _omitting_ all of them seems dangerous for me....

Reply:
By removing the `instanceof` check (which I don't think is strictly necessary) and using optional chaining there's the potential to turn 100 lines of code to 10 lines of code:

**Before**
```js
// Given an object that was obtained via Scene#pick, examine it
// to see whether it contains tile metadata.
// If the object contains TileMetadata, then this metadata
// is returned.
// Otherwise, 'undefined' is returned.
const obtainTileMetadata = function (picked) {
  const metadata = picked?.content?.tile?.metadata;
  const isTileMetadata = metadata instanceof Cesium.TileMetadata;
  if (!isTileMetadata) {
    return undefined;
  }
  return metadata;
};

// Given an object that was obtained via Scene#pick, examine it
// to see whether it contains tileset metadata.
// If the object contains TilesetMetadata, then this metadata
// is returned.
// Otherwise, 'undefined' is returned.
const obtainTilesetMetadata = function (picked) {
  if (!Cesium.defined(picked)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content.tileset)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content.tileset.metadata)) {
    return undefined;
  }
  const metadata = picked.content.tileset.metadata.tileset;
  const isTilesetMetadata = metadata instanceof Cesium.TilesetMetadata;
  if (!isTilesetMetadata) {
    return undefined;
  }
  return metadata;
};

// Given an object that was obtained via Scene#pick, examine it
// to see whether it contains group metadata.
// If the object contains GroupMetadata, then this metadata
// is returned.
// Otherwise, 'undefined' is returned.
const obtainGroupMetadata = function (picked) {
  if (!Cesium.defined(picked)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content.groupMetadata)) {
    return undefined;
  }
  const metadata = picked.content.groupMetadata;
  const isGroupMetadata = metadata instanceof Cesium.GroupMetadata;
  if (!isGroupMetadata) {
    return undefined;
  }
  return metadata;
};

// Given an object that was obtained via Scene#pick, examine it
// to see whether it contains content metadata.
// If the object contains ContentMetadata, then this metadata
// is returned.
// Otherwise, 'undefined' is returned.
const obtainContentMetadata = function (picked) {
  if (!Cesium.defined(picked)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content.metadata)) {
    return undefined;
  }
  const metadata = picked.content.metadata;
  const isContentMetadata = metadata instanceof Cesium.ContentMetadata;
  if (!isContentMetadata) {
    return undefined;
  }
  return metadata;
};

// Install the handler that will check the element that is
// under the mouse cursor when the mouse is moved, and
// add any metadata that it contains to the label.
const handler = new Cesium.ScreenSpaceEventHandler(viewer.scene.canvas);
handler.setInputAction(function (movement) {
  labelEntity.label.text = "";
  const picked = viewer.scene.pick(movement.endPosition);

  const tilesetMetadata = obtainTilesetMetadata(picked);
  addMetadataToLabel("Tileset metadata", tilesetMetadata);

  const tileMetadata = obtainTileMetadata(picked);
  addMetadataToLabel("Tile metadata", tileMetadata);

  const groupMetadata = obtainGroupMetadata(picked);
  addMetadataToLabel("Group metadata", groupMetadata);

  const contentMetadata = obtainContentMetadata(picked);
  addMetadataToLabel("Content metadata", contentMetadata);

  const cartesian = viewer.scene.pickPosition(movement.endPosition);
  labelEntity.position = cartesian;
}, Cesium.ScreenSpaceEventType.MOUSE_MOVE);
```

**After**
```js
// Install the handler that will check the element that is
// under the mouse cursor when the mouse is moved, and
// add any metadata that it contains to the label.
const handler = new Cesium.ScreenSpaceEventHandler(viewer.scene.canvas);
handler.setInputAction(function (movement) {
  labelEntity.label.text = "";
  const picked = viewer.scene.pick(movement.endPosition);

  const tilesetMetadata = picked?.content?.tileset?.metadata?.tileset;
  const tileMetadata = picked?.content?.tile?.metadata;
  const groupMetadata = picked?.content?.groupMetadata;
  const contentMetadata = picked?.content?.metadata;
  
  addMetadataToLabel("Tileset metadata", tilesetMetadata);
  addMetadataToLabel("Tile metadata", tileMetadata);
  addMetadataToLabel("Group metadata", groupMetadata);
  addMetadataToLabel("Content metadata", contentMetadata);

  const cartesian = viewer.scene.pickPosition(movement.endPosition);
  labelEntity.position = cartesian;
}, Cesium.ScreenSpaceEventType.MOUSE_MOVE);
```

Reply:
This is also revealing some asymmetry in the API.

```js
  const tilesetMetadata = picked?.content?.tileset?.metadata?.tileset;
  const tileMetadata = picked?.content?.tile?.metadata;
  const groupMetadata = picked?.content?.groupMetadata;
  const contentMetadata = picked?.content?.metadata;
```

Instead I would expect the code below.

```js
  const tilesetMetadata = picked?.content?.tileset?.metadata;
  const tileMetadata = picked?.content?.tile?.metadata;
  const groupMetadata = picked?.content?.group?.metadata;
  const contentMetadata = picked?.content?.metadata;
```

@ptrgags @j9liu is this fixable?

Reply:
@lilleyse yes, I think this is pretty easy:

* for the tileset case, tha `tileset.metadata` object is currently a `Cesium3DTilesetMetadata` object which contains the schema, list of all group metadata objects, and the tileset metadata. If we make that object a private field of `Cesium3DTileset` and replace it with getters `tileset.metadata` (which will now return a `TilesetMetadata` object), `tileset.schema` and `tileset.groups`, these can just delegate to that object. This way, we don't have to clutter `Cesium3DTileset` with more responsibilities (it already does far too much)
* for the group metadata, we just need to add a simple wrapper object of a `Group` that has a `metadata` field for consistency of API.

Reply:
I opened https://github.com/CesiumGS/cesium/pull/10224 for this.

Reply:
And now merged. @javagl could you update to the new API?